### PR TITLE
Kokkos launch compiler updates

### PR DESCRIPTION
**Pull request by jrmadsen**
- Creates new `find_package(Kokkos COMPONENTS launch_compiler)` mode for using the launch compiler feature regardless of whether CUDA is enabled
- `kokkos_launch_compiler` was extended to support forwarding to the original compiler Kokkos was compiled with
- There is only one known caveat: kokkos is installed with `nvcc_wrapper` and downstream application uses clang
  - This results in the default host compiler to use NVCC_WRAPPER_DEFAULT_COMPILER from environment or nothing and without specify NVCC_WRAPPER_DEFAULT_COMPILER during compilation clang will be used and it will fail
  - Moral of the story: don't use nvcc_wrapper anymore when installing kokkos

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -144,8 +144,12 @@ GLOBAL_SET(KOKKOS_XCOMPILER_OPTIONS)
 # We need to append text here for making sure TPLs
 # we import are available for an installed Kokkos
 GLOBAL_SET(KOKKOS_TPL_EXPORTS)
-# this could probably be scoped to project
+# KOKKOS_DEPENDENCE is used by kokkos_launch_compiler
 GLOBAL_SET(KOKKOS_COMPILE_DEFINITIONS KOKKOS_DEPENDENCE)
+# MSVC never goes through kokkos_launch_compiler
+IF(NOT MSVC)
+    GLOBAL_APPEND(KOKKOS_LINK_OPTIONS -DKOKKOS_DEPENDENCE)
+ENDIF()
```
Comment (masterleinad) [inline CMakeLists.txt:152]: Can please again elaborate on the issue this is supposed to fix?

File: bin/kokkos_launch_compiler [hunk 2]
```diff
@@ -23,65 +34,88 @@ do
     fi
 done
 
-# if C++ is not passed, someone is probably trying to invoke it directly
+# if Kokkos compiler is not passed, someone is probably trying to invoke it directly
 if [ -z "${1}" ]; then
-    echo -e "\n${BASH_SOURCE[0]} was invoked without the C++ compiler as the first argument."
+    echo -e "\n${BASH_SOURCE[0]} was invoked without the Kokkos compiler as the first argument."
+    echo "This script is not indended to be directly invoked by any mechanism other"
+    echo -e "than through a RULE_LAUNCH_COMPILE or RULE_LAUNCH_LINK property set in CMake\n"
```
Comment (masterleinad) [inline bin/kokkos_launch_compiler:41]:
```suggestion
    echo -e "than through a RULE_LAUNCH_COMPILE or RULE_LAUNCH_LINK property set in CMake.\n"
```

File: bin/kokkos_launch_compiler [hunk 3]
```diff
@@ -23,65 +34,88 @@ do
...
+    exit 1
+fi
+
+# if Kokkos compiler is not passed, someone is probably trying to invoke it directly
+if [ -z "${2}" ]; then
+    echo -e "\n${BASH_SOURCE[0]} was invoked without the C++ compiler as the second argument."
     echo "This script is not indended to be directly invoked by any mechanism other"
     echo -e "than through a RULE_LAUNCH_COMPILE or RULE_LAUNCH_LINK property set in CMake\n"
```
Comment (masterleinad) [inline bin/kokkos_launch_compiler:49]:
```suggestion
    echo -e "than through a RULE_LAUNCH_COMPILE or RULE_LAUNCH_LINK property set in CMake.\n"
```

File: CMakeLists.txt (hunk 1 above)
Comment (jrmadsen) [reply CMakeLists.txt:152]: Well A, that flag wasn't showing up in the exported link flags. And B, the new system should always have that present, not just when CUDA is used.

Comment (jrmadsen) [reply CMakeLists.txt:152]: I believe the reason it wasn't showing up in the exported link flags might have been that it was getting added to the `KOKKOS_LINK_OPTIONS` after `KOKKOS_LINK_OPTIONS` was exported, so it was getting applied in the build tree but the install tree export didn't have it. Didn't dive into too much bc re-adding it here is basically what we need to do anyway to make sure it always there.

**Comment by jrmadsen** [comment]
Retest this please

**Comment by jrmadsen** [comment]
Retest this please

**Comment by crtrott** [comment]
I guess this is ok? But I just don't know. What happened before if you installed Kokkos with nvcc_wrapper and then downstream use clang? But we can just say we don't support that.

**Comment by jrmadsen** [comment]
> What happened before if you installed Kokkos with nvcc_wrapper and then downstream use clang?

Before this, i don't know. You probably just got compiler-flag errors since Kokkos would export compiler flags for NVCC. After this PR, the problem with installing Kokkos via nvcc_wrapper is that we don't actually know the underlying host compiler at CMake time (since they could configure and then set `NVCC_WRAPPER_DEFAULT_COMPILER` in the env later) so it can't save that compiler for downstream projects. Thus if the default `g++` is version 4.9 and the user set `NVCC_WRAPPER_DEFAULT_COMPILER=g++-8` when building Kokkos after cmake was run, kokkos_launch_compiler in the install tree would not "know" to set `NVCC_WRAPPER_DEFAULT_COMPILER` to g++-8.

**Comment by jrmadsen** [comment]
I actually think this makes a rather strong argument for us deprecating the use of `nvcc_wrapper`. E.g. make a really annoying warning in the Kokkos CMake that says stop setting the C++ compiler to nvcc_wrapper.

**Comment by jrmadsen** [comment]
And it would be a win in my and @DavidPoliakoff war against compiler-wrappers.

**Comment by jrmadsen** [comment]
> But we can just say we don't support that.

Also, now _it is supported_. Since you could do `find_package(Kokkos COMPONENTS launch_compiler)` and even if the downstream compiler were clang, you would end up compiling your kokkos code with nvcc_wrapper.

**Review by DavidPoliakoff — APPROVED** [review]

**Pull request merged**